datalad/support/gitrepo.py

```diff
@@ -3100,6 +3069,82 @@ def _get_cache_key(label, paths, ref, untracked=None):
         else:
             return status
 
+    def _get_worktree_modifications(self, paths=None):
+        """Report working tree modifications
+
+        Parameters
+        ----------
+        paths : list or None
+          If given, limits the query to the specified paths. To query all
+          paths specify `None`, not an empty list.
+
+        Returns
+        -------
+        dict
+          Mapping of modified Paths to type labels from GIT_MODE_TYPE_MAP.
+          Deleted paths have type `None` assigned.
+        """
+        # because of the way git considers smudge filters in modification
+        # detection we have to consult two commands to get a full picture, see
+        # https://github.com/datalad/datalad/issues/6791#issuecomment-1193145967
+
+        # low-level code cannot handle pathobjs
+        consider_paths = [str(p) for p in paths] if paths else None
+
+        # first ask diff-files which can report typechanges. it gives a list with
+        # interspersed diff info and filenames
+        mod = list(self.call_git_items_(
+            ['diff-files',
+             # without this, diff-files would run a full status (recursively)
+             # but we are at most interested in a subproject commit
+             # change within the scope of this repo
+             '--ignore-submodules=dirty',
+             # hopefully making things faster by turning off features
+             # we would not benefit from (at least for now)
+             '--no-renames',
+             '-z'
+            ],
+            files=consider_paths, sep='\0', read_only=True))
+        # convert into a mapping path to type
+        modified = dict(zip(
+            # paths are every other element, starting from the second
+            mod[1::2],
+            # mark `None` for deletions, and take mode reports otherwise
+            # (for simplicity keep leading ':' in prev mode for now)
+            (None if spec.endswith('D') else spec.split(' ', maxsplit=2)[:2]
+             for spec in mod[::2])
+        ))
+        # `diff-files` cannot give us the full answer to "what is modified"
+        # because it won't consider what smudge filters could do, for this
+        # we need `ls-files --modified` to exclude any paths that are not
+        # actually modified
+        modified_files = set(
+            p for p in self.call_git_items_(
+                # we need not look for deleted files, diff-files did that
+                ['ls-files', '-z', '-m'],
```

Comment:
possible enhancement to do since there was a concern of now having two calls instead of one:  this call (`ls-files`) seems to be independent of prior `diff-files`.  . Could at least partially be mitigated by multi-threading first execution until the 2nd one is complete (AKA poor man "async" ;-) ).  We have `parallel.ProducerConsumer` but I think it would be too heavy for such one.  It is more of really a simply do it directly here without bothering with a helper until we have another use case like this.